src/core/account_builder.cpp

```diff
@@ -0,0 +1,278 @@
+/* 
+File: account_builder.cpp
+
+Brief:  Account builder performs exception handling and guides the user
+        through the process of creating an account.
+
+Authors: Daniel Mendez 
+         Alex Salazar
+         Arman Alauizadeh 
+         Alexander DuPree
+         Kyle Zalewski
+         Dominique Moore
+
+https://github.com/AlexanderJDupree/ChocAn
+ 
+*/
+
+#include <ChocAn/core/utils/account_builder.hpp>
+#include <ChocAn/core/utils/validators.hpp>
+#include <algorithm>
+
+
+bool Account_Builder::buildable() const
+{
+    return build_phase.empty() && account_accepted;
+}
+
+Account Account_Builder::build()
+{
+    invalid_account_build errors("",{});
+
+    if (!buildable())
+        throw invalid_account_build(errors.default_msg, {errors.building_unbuildable});
+
+    if (account_field.at(current_type) == "Manager")
+        return Account(name.value(), address.value(), Manager(), id_generator);
+
+    if (account_field.at(current_type) == "Provider")
+        return Account(name.value(), address.value(), Provider(), id_generator);
+    
+    return Account(name.value(), address.value(), Member(), id_generator);
+
+}
+
+Account_Builder &Account_Builder::reset()
+{
+
+    while (!build_phase.empty())
+        build_phase.pop();
+
+    Build_Instructions building_phase;
+    
+    build_phase.push(building_phase.street_address);
+    build_phase.push(building_phase.full_name);
+    build_phase.push(building_phase.account_type);
+
+    name.reset();
+    address.reset();
+    zip.reset();
+    issues.reset();
+
+    account_accepted = false;
+
+    return *this;
+}
+
+void Account_Builder::set_current_field(const std::string &input)
+{
+
+    Build_Instructions building_phase;
+   
+    if (build_phase.empty())
+    {
+
+        account_accepted = confirm_account(input);
+
+        if (!account_accepted){
+
+            reset();
+            issues.emplace(invalid_account_build(building_phase.account_rejected,{ "With input value: " + input}));
+
+        }
+        
+        return;
+    }
+
+    std::string current_phase = build_phase.top();
+
+    if (current_phase == building_phase.account_type)
```

Comment:
Essentially you're using the `Build_Instructions` string members in a very expensive switch statement. If we get rid of the strings and replace them with a variant or enumeration we don't have to do a 30 character something string compare on each else if and just run the right logic.

Reply:
im going to implement a variant to track build phase. the Build_Instructions struct will be gone and with it any dependencies. each variant will have its own string message to handle input instructions. no conditionals will be needed then as the variance in output will be handled by the variant.

Reply:
the state of the build process in handled by a stack of variants, where the type held represents the build state. an account is build able _iff_ this stack is empty. this allows the logic for getting inputs from the user for use as account fields to be used in both creating and updating account states. the stack is managed such that it will only pop if the most recent input did not cause any exceptions. this locks the account builder in its current state until a valid input is recieved. the validation is performed at the entity level by attempting to construct them and allowing there smart constructors to determine whether or not the input is valid. when creating and account, valid entities are maintained throughout the entire process, with reconstruction attempts made every time new input is received. once buildable, an account will be constructed with these entities. when updating an account the input validation is handled the same. the only difference is instead of building a new account with these entities, an account refrence is passed in for applying updates to. the account will reconstruct its entities with either a value from the account builder or the value currently held by the entity itself.